File: src-tauri/src/api/ai_api.rs
use std::sync::Arc;

use brainy_core::{
    Guid,
    ai_integration::{
        ai_service::{AiService, StreamLlmResponseEvent},
        ai_state::AiState,
        entities::{chat::Chat, message::Message},
    },
    common::traits::repositories_context::RepositoriesContext,
};
use tauri::{State, ipc::Channel};
use tokio::sync::Mutex;

use crate::api::ApiError;

#[tauri::command]
pub async fn stream_ai_response(
    context: State<'_, Arc<Mutex<dyn RepositoriesContext>>>,
    ai_service: State<'_, Arc<AiService>>,
    on_event: Channel<StreamLlmResponseEvent>,
    prompt: String,
    chat_id: Option<Guid>,
) -> Result<(), ApiError> {
    let result = ai_service
        .stream(prompt, chat_id, |event| match on_event.send(event) {
            Ok(_) => Ok(()),
            Err(err) => Err(err.to_string()),
        })
        .await;
    let context = context.lock().await;
    context.save_changes().await?;

    match result {
        Ok(()) => Ok(()),
        Err(err) => Err(ApiError::new(err.to_string())),
    }
}

#[tauri::command]
pub async fn stop_ai_generation(ai_state: State<'_, Arc<AiState>>) -> Result<(), ApiError> {
    ai_state.cancel_generation();
    Ok(())
}

#[tauri::command]
pub async fn get_all_ai_chats_sorted_by_date_desc(
    context: State<'_, Arc<Mutex<dyn RepositoriesContext>>>,
) -> Result<Vec<Chat>, ApiError> {
    let context = context.lock().await;
    let chats = context
        .ai_repository()
        .get_all_chats_sorted_by_date_desc()
        .await?;
    Ok(chats)
}

#[tauri::command]
pub async fn delete_ai_chat(
    context: State<'_, Arc<Mutex<dyn RepositoriesContext>>>,
    id: Guid,
) -> Result<(), ApiError> {
    let context = context.lock().await;
    context.ai_repository().delete_chat(id).await?;
    context.save_changes().await?;
    Ok(())
}

#[tauri::command]
pub async fn get_chat_messages_ordered(
    context: State<'_, Arc<Mutex<dyn RepositoriesContext>>>,
    id: Guid,
) -> Result<Vec<Message>, ApiError> {
    let context = context.lock().await;
    let messages = context
        .ai_repository()
        .get_chat_messages_ordered(id)
        .await?;
    Ok(messages)
}

File: src/api/aiIntegration/api/aiApi.ts
import { Channel, invoke } from "@tauri-apps/api/core";
import { StreamLlmResponseEvent } from "../events/streamLlmResponseEvent";
import Chat from "../entities/chat";
import Message from "../entities/message";
import StreamAiRequestDto from "../dto/streamAiRequestDto";

export function streamAiResponse(
	request: StreamAiRequestDto,
	onEvent: Channel<StreamLlmResponseEvent>,
): Promise<void> {
	return invoke("stream_ai_response", {
		request,
		onEvent,
	});
}

export function stopAiGeneration(): Promise<void> {
	return invoke("stop_ai_generation");
}

export function suggestClozeContent(content: string): Promise<string> {
	return invoke("suggest_cloze_content", { content });
}

export function getAllAiChatsSortedByDateDesc(): Promise<Chat[]> {
	return invoke("get_all_ai_chats_sorted_by_date_desc");
}

export function deleteAiChat(id: string): Promise<string> {
	return invoke("delete_ai_chat", { id });
}

export function getChatMessagesOrdered(id: string): Promise<Message[]> {
	return invoke("get_chat_messages_ordered", { id });
}

export function renameAiChat(id: string, newTitle: string): Promise<void> {
	return invoke("rename_ai_chat", { id, newTitle });
}

export function rejectToolCall(messageId: string): Promise<void> {
	return invoke("reject_tool_call", { messageId });
}

export function acceptToolCall(messageId: string): Promise<void> {
	return invoke("accept_tool_call", { messageId });
}

export function uploadDocument(path: string, chatId: string): Promise<void> {
	return invoke("upload_document", { path, chatId });
}
